server/src/sas/CompletionProvider.ts

```diff
@@ -338,6 +343,406 @@ export class CompletionProvider {
+      const expectedCurLineIndentText = this._makeIndentText(
+        expectedCurLineIndent,
+        useSpace,
+        tabSize,
+      );
+      return [
+        TextEdit.replace(
+          {
+            start: { line: line, character: 0 },
+            end: { line: line, character: curLineIndentText.length },
+          },
+          expectedCurLineIndentText,
+        ),
+      ];
+    }
+  }
+
+  private _getEnterTriggeredIndentEdit(
+    line: number,
+    tabSize: number,
+    useSpace: boolean,
+  ): TextEdit[] {
+    // find indent text
+    const lastNotEmptyLine = this._getLastNotEmptyLine(line);
+    if (lastNotEmptyLine === undefined) {
+      return [];
+    }
+    const lastNotEmptyLineText = this.model.getLine(lastNotEmptyLine);
+    const nextLineText =
+      line < this.model.getLineCount() - 1 ? this.model.getLine(line + 1) : "";
+    const nextLineIndentText = this._getIndentText(nextLineText);
+
+    // calculate indent length
+    const curIndentLen = this._getIndentLength(lastNotEmptyLineText, tabSize);
+    const nextLineIndentInc: number | undefined =
+      this._getIndentIncrementOfNextLine(
+        lastNotEmptyLine,
+        curIndentLen,
+        tabSize,
+      );
+    if (nextLineIndentInc === undefined) {
+      return [];
+    }
+    const expectedNextLineIndentLen = curIndentLen + nextLineIndentInc;
+    const actualNextLineIndentLen = this._getIndentLength(
+      nextLineText,
+      tabSize,
+    );
+
+    if (expectedNextLineIndentLen === actualNextLineIndentLen) {
+      return [];
+    } else {
+      const expectedNextLineIndentText = this._makeIndentText(
+        expectedNextLineIndentLen,
+        useSpace,
+        tabSize,
+      );
+      return [
+        TextEdit.replace(
+          {
+            start: { line: line + 1, character: 0 },
+            end: { line: line + 1, character: nextLineIndentText.length },
+          },
+          expectedNextLineIndentText,
+        ),
+      ];
+    }
+  }
+
+  private _getIndentIncrementOfNextLine(
+    line: number,
+    curIndent: number,
+    tabSize: number,
+  ): number | undefined {
+    // find semicolon token
+    const tokens: SyntaxToken[] = this.syntaxProvider.getSyntax(line);
+    const cleanedTokens = this._cleanTokens(line, tokens);
+    if (cleanedTokens.length === 0) {
+      return 0;
+    }
+    // find patterns of "data xxx;", "proc xxx;" or "%macro xxx;"
+    const lineText = this.model.getLine(line);
+    let curIndex = cleanedTokens.length;
+    do {
+      curIndex = this._findSemicolonTokenRightToLeft(
+        line,
+        cleanedTokens,
+        curIndex - 1,
+      );
+      if (curIndex === -1) {
+        return 0;
+      }
+      const semicolonToken: SyntaxToken = cleanedTokens[curIndex];
+      const prevToken = cleanedTokens[curIndex - 1]; // curIndex must be > 0
+      const prevTokenText = this._getTokenText(
+        cleanedTokens,
+        curIndex - 1,
+        lineText,
+      )?.trim();
+      if (
```

Comment:
Code zone of "run", "quit" have no different from other statements. If I just get the zone of "stmt", and "run":
data xxx; 
    stmt;
run;
They all return DATA_STEP_STMT: 540, I can't distinguish them.

But I can add an additional check of token.style: In addition to be "run","quit","%mend", they must also be "sec-keyword".

Reply:
for section boundary keywords, zone before and after semicolon should be different, right?
sec-keyword may have recursive section problem, for example "%macro;data;run;"

Reply:
I tried what you say, but the code zone is not always reliable for this, especially when comments occur. For example:
`data a;run/*abc*/;`
Code zone before semicolon is DATA_STEP_STMT_OPT: 541, and after semicolon is undefined.

`data a;run;/*abc*/`
Code zone before semicolon is GBL_STMT: 500, and after semicolon is GBL_STMT: 500.

There may be more error cases, cases I provide may not be all.

**sec-keyword may have recursive section problem, for example "%macro;data;run;"**
I think this is another topic, whether we use code zone or not, we can't perfectly resolve this problem. We need to locate the FoldingBlock of the data section(because data; may last for many lines, not only one line). As we don't support recursive FoldingBlock now, we can't resolve it.